engine/src/main/java/org/terasology/engine/rendering/assets/mesh/resource/IndexResource.java

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.engine.rendering.assets.mesh.resource;
+
+import org.lwjgl.BufferUtils;
+import java.nio.ByteBuffer;
+
+/**
+ * defines the order of vertices to walk for rendering geometry
+ *
+ * refrence: https://www.khronos.org/opengl/wiki/Primitive
+ */
+public class IndexResource {
+    public ByteBuffer buffer;
+    private int numIndices = 0;
+    private int inSize = 0;
+    private int posIndex = 0;
+
+    public int getNumberOfIndices() {
+        return numIndices;
+    }
+
+    public int getSize() {
+        return numIndices * Integer.BYTES;
+    }
+
+    public IndexResource() {
+        this.buffer = BufferUtils.createByteBuffer(0);
+    }
+
+    public void ensureCapacity(int size) {
+        if (size > buffer.capacity()) {
+            int newCap = Math.max(this.inSize << 1, size);
+            ByteBuffer newBuffer = BufferUtils.createByteBuffer(newCap);
+            buffer.limit(Math.min(size, inSize));
+            buffer.position(0);
+            newBuffer.put(buffer);
+            this.buffer = newBuffer;
+        }
+        if (size > this.inSize) {
+            this.inSize = size;
+        }
+    }
+
+    public void reserveElements(int elements) {
+        ensureCapacity(elements * Integer.BYTES);
+    }
+
+    public void rewind() {
+        posIndex = 0;
+    }
+
+    public void put(int value) {
+        ensureCapacity((posIndex + 1) * Integer.BYTES);
```

Comment:
Is it possible for this function to be called multiple times, say in a loop? If yes, then wouldn't it make more sense to not put it here/have an alternate overload that takes a list and only performs the check once for the entire list?

Reply:
the check should be quick. the buffer is doubled each time you overflow its content. its dynamic in how the data is getting allocated.